frame/safe-mode/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+
+		type EnableOrigin: EnsureOrigin<Self::Origin>;
+		type ExtendOrigin: EnsureOrigin<Self::Origin>;
+		type PreemptiveDisableOrigin: EnsureOrigin<Self::Origin>;
+
+		// Weight information for extrinsics in this pallet.
+		//type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The safe-mode is (already) enabled.
+		IsEnabled,
+
+		/// The safe-mode is (already) disabled.
+		IsDisabled,
+
+		/// No permission to disable the safe-mode (yet).
+		NoPermission,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The safe-mode was enabled until inclusively this block.
+		Enabled(T::BlockNumber),
+
+		/// The safe-mode was extended until inclusively this block.
+		Extended(T::BlockNumber),
+
+		/// The safe-mode was disabled.
+		Disabled,
+	}
+
+	/// Contains the last block number that the safe-mode will stay enabled.
+	///
+	/// This is set to `None` if the safe-mode is disabled.
+	/// In any block after this block number is reached,
+	/// the safe-mode can be disabled permissionlessly by any-one.
+	#[pallet::storage]
+	pub type Enabled<T: Config> = StorageValue<_, T::BlockNumber, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Enable the safe-mode.
+		#[pallet::weight(0)]
+		pub fn enable(origin: OriginFor<T>) -> DispatchResult {
+			T::EnableOrigin::ensure_origin(origin)?;
+
+			ensure!(!Enabled::<T>::exists(), Error::<T>::IsEnabled);
+			let limit =
+				<frame_system::Pallet<T>>::block_number().saturating_add(T::EnableDuration::get());
+			Enabled::<T>::put(limit);
+			Self::deposit_event(Event::Enabled(limit));
+
+			Ok(())
+		}
+
+		/// Extends the safe-mode duration.
+		///
+		/// Can only be called by the [`Config::ExtendOrigin`] origin.
+		/// The safe-mode must already be enabled for this to succeed.
+		///
+		/// NOTE: This extends the *original* safe-mode duration.
+		/// The extension period can be less than [`Config::ExtendDuration`]
+		/// if the safe-mode period already ran out.
+		/// So to say: even a delayed call to `extend` will never result in
+		/// an extension period of more than [`Config::ExtendDuration`] blocks.
+		#[pallet::weight(0)]
+		pub fn extend(origin: OriginFor<T>) -> DispatchResult {
+			T::ExtendOrigin::ensure_origin(origin)?;
+
+			let limit = Enabled::<T>::take()
+				.ok_or(Error::<T>::IsDisabled)?
+				.saturating_add(T::ExtendDuration::get());
+			Enabled::<T>::put(limit);
+			Self::deposit_event(Event::<T>::Extended(limit));
+
+			Ok(())
+		}
+
+		/// Disable the safe-mode.
+		///
+		/// Can be called either by
+		///  [`Config::PreemptiveDisableOrigin`] at any time
+		/// or
+		///	 anyone after block number [`Enabled`]+1 is reached.
+		#[pallet::weight(0)]
+		pub fn disable(origin: OriginFor<T>) -> DispatchResult {
+			let can_preempt = if T::PreemptiveDisableOrigin::ensure_origin(origin.clone()).is_ok() {
+				true
+			} else {
+				ensure_signed(origin)?;
+				false
+			};
```

Comment:
How do we make sure this is something that is always callable?

Sounds like we should have scheduler trigger this automatically, to prevent any foot guns.

Reply:
Yes should probably be automatic.  
However, the scheduler pallet is not *guaranteed* to execute a call, but we could do it in `on_initialize`.  
I think it makes sense to keep the extrinsic as preemptive disable though.

Reply:
Yes, keeping the extrinsic is no problem, and I believe an extrinsic would be needed anyway if we used the scheduler, but we should make it somehow by default that something would automatically unlock the chain after some period of time.

Reply:
I added it to the `on_initialize` since we read the `Enabled` field at least once anyway, so it should be in the overlay.